packages/react/src/components/Layout/index.tsx

```diff
@@ -0,0 +1,224 @@
+
+export interface LayoutConstraintProps extends HTMLAttributes<HTMLElement> {
+  /**
+   * Specify a custom component or element to be rendered as the top-level
+   * element in the component
+   */
+  as?: (() => ReactNode) | string | ElementType;
+
+  /**
+   * Provide child elements to be rendered inside of `LayoutConstraint`
+   */
+  children?: ReactNode;
+
+  /**
+   * Provide a custom class name to be used on the outermost element rendered by
+   * the component
+   */
+  className?: string;
+
+  /**
+   * Specify the desired layout density constraints of this element's children
+   */
+  density?: {
+    min?: Density | null;
+    default?: Density | null;
+    max?: Density | null;
+  };
+
+  /**
+   * Specify the desired layout size constraints of this element's children
+   */
+  size?: {
+    min?: Size;
+    default?: Size;
+    max?: Size;
+  };
+}
+
+const LayoutConstraint = React.forwardRef<ReactNode, LayoutConstraintProps>(
+  function Layout(
+    { as: BaseComponent = 'div', children, className, density, size, ...rest },
+    forwardRef
+  ) {
+    const prefix = usePrefix();
+
+    const classes = cx(
+      className,
+      Object.entries({
+        size,
+        density,
+      }).map(([group, constraints]) => ({
+        [`${prefix}--layout-constraint--${group}:default-${constraints?.default}`]:
+          constraints?.default,
+        [`${prefix}--layout-constraint--${group}:min-${constraints?.min}`]:
+          constraints?.min,
+        [`${prefix}--layout-constraint--${group}:max-${constraints?.max}`]:
+          constraints?.max,
+      }))
+    );
+
+    return (
+      <BaseComponent {...rest} ref={forwardRef} className={classes}>
+        {children}
+      </BaseComponent>
+    );
+  }
+);
+
+LayoutConstraint.propTypes = {
+  /**
+   * Specify a custom component or element to be rendered as the top-level
+   * element in the component
+   */
+  as: PropTypes.oneOfType([
+    PropTypes.func,
+    PropTypes.string,
+    PropTypes.elementType,
+  ]),
+
+  /**
+   * Provide child elements to be rendered inside of `LayoutConstraint`
+   */
+  children: PropTypes.node,
+
+  /**
+   * Provide a custom class name to be used on the outermost element rendered by
+   * the component
+   */
+  className: PropTypes.string,
+
+  /**
+   * Specify the desired layout density constraints of this element's children
+   */
+  density: PropTypes.shape({
+    min: PropTypes.oneOf(['condensed', 'normal']),
+    default: PropTypes.oneOf(['condensed', 'normal']),
+    max: PropTypes.oneOf(['condensed', 'normal']),
+  }),
+
+  /**
+   * Specify the desired layout size constraints of this element's children
+   */
+  size: PropTypes.shape({
+    min: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl', '2xl']),
+    default: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl', '2xl']),
+    max: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl', '2xl']),
+  }),
+} as WeakValidationMap<
+  LayoutConstraintProps & React.RefAttributes<React.ReactNode>
+>;
```

Comment:
This conversion shouldn't be necessary with the suggested changes
```suggestion
}
```